### ISSUE TITLE
Do we need Patterns?

### ISSUE THREAD

**Issue by ericwbailey** [a11yproject/a11yproject.com#721]
It's tough to know what the end context will be for an implemented pattern, so my question is does this page do more harm than good if people are just copy/pasting the content?

Maybe we could rework it to something like Dave's [a11y nutrition cards](https://davatron5000.github.io/a11y-nutrition-cards/), where we provide guidance for the expected behaviors for the major UI components.

If we do keep our Patterns page, we should be sure to audit to verify that all patterns work as expected across all major assistive tech.

**Comment by scottaohara** [comment]
I could see us keeping a pattern page, or a version of it. But I think explaining potential usage would help in some cases. Why is this pattern accessible, how should it be used. How could misuse make for a poor ux?

Definitely a lot to work with here, but as you said, we should audit and think on this some more.

**Comment by RyanAaronGreen** [comment]
My two cents - I know that I've benefited greatly from having patterns to look at when learning how to go about making a section accessible. Especially in the beginning when I didn't even know how to go about using javascript very well. 

With that being said, maybe we can switch the focus of what's more important? Perhaps we can merge the cards AND the patterns into one thing where the focus is on the "why/how" and then show the patterns as examples of that below. 

For example, there could be a list page with all complete _modules/widgets_. Once you click on a _module/widget_, you're taken to another page where it dives more in depth on how to make that particular _module/widget_ accessible (i.e. following something similar to Dave's work) - then below that are the patterns of what that can look like in code.

That way people will be inclined to read the why & how and then be able to dig into examples of how it was done. It also creates less of a need to make sure that the patterns work in every possible scenario as it's only there to teach, not necessarily to be copied and used.

**Comment by scottaohara** [comment]
been thinking of this some more lately and i do like Ryan's take here.

the patterns by themselves are neat...but problematic in that there is little in the way of reasoning or indication if the pattern is still appropriate with the ever changing technology landscape.

I think we should be mindful about the patterns we choose to showcase and how we present them in the redesign.

**Comment by ericwbailey** [comment]
More ammo for distancing ourselves from actual implementations: https://github.com/a11yproject/a11yproject.com/issues/803

**Comment by golgistudio** [comment]
Just stumbled upon this site.  Very helpful!

I like the idea of the patterns, but agree that the current implementation could be improved.

The Salesforce patterns is along the lines of the nutrition cards mentioned above, but includes examples.
https://lightningdesignsystem.com/accessibility/patterns/accordion/

Here's another example
https://www.accessibility-developer-guide.com/examples/hiding-elements/visually/

My thoughts on the patterns section...

I think examples are important and being able to run them on something like codepen.io provides a good starting point for understanding the interaction.

Just because the example doesn't fully address all accessibility requirements doesn't necessarily mean it is not helpful.  Adding a qualification when an example only demonstrates one aspect of accessibility will provide the user some context.

Pattern pages can be marked as draft or reviewed to further indicate completeness.

**Comment by SebacicA11y** [comment]
I'm currently playing with Accordions, and the patterns available on The A11y Project are very helpful. But there are links to Scott O'Hara's GitHub, and he then links to ARIA authoring practices. These have different models and "requirements".  Reading the actual WCAG guidelines shows that Accordions require implementation across all four POUR principles. 

As mentioned above, cut and paste from patterns is not going to work. But they are a great starting point. Without them, I for one would never learn anything. 

I would like to see a DIRECT reference to the POUR principles, so that I can make an "informed" decision about which aspects of each of the patterns I "have" to implement and which ones will work best with my design. 

For example, I am currently struggling to decide if Accordions "have" to have <Headings> or if <Buttons> are OK.

**Comment by scottaohara** [comment]
@SebacicA11y this is exactly one of the issues with having these patterns here. There are different patterns which may be appropriate based on the content and desired ux of a document. Also, depending on which accordion pattern you are referring to, there are patterns which may be dated, in that they were appropriate patterns at the time they were created, but have fallen out of favor.

Taking Dave’s cards as an example, they reference using aria-controls to point from the accordion’s trigger to panel. However, the only screen reader that supports aria-controls was JAWS, and it recently set aria-controls to be ignored by default, because usage like this received feed back as being overly verbose/unnecessary from users

**Comment by SebacicA11y** [comment]
Interesting. So do Jaws and the other screen readers have suggested
patterns? I would have thought that it would be in their best interests to
clearly explain what works best with them.

**Comment by scottaohara** [comment]
Screen readers and browsers implement support for ARIA based on the ARIA spec.  

ARIA authoring practice guidelines (apg) document suggested methods to implement certain widgets / common components, but APG is not a specification and deviations can be made from these patterns if testing / feedback prove alternate configurations may work better.

In regards to `aria-controls`, there is no language in the specification to dictate what screen readers should or must allow for, if used.  JAWS had support for this feature, announcing "use jaws key + alt + m to move to controlled element" when it was encountered.  However, this was not found to be useful in most situations it was used in, as the controlled element was often the next element in the DOM, or Tab focus order.  Hence JAWS changing default settings to ignore the attribute unless someone specifically turns the functionality on.

**Comment by SebacicA11y** [comment]
OK. 
So, to my mind we need to "leverage a worldwide community of developer knowledge" and make patterns available from that community. Sure things will change, and people will add to the knowledge. But without the patterns I wouldn't be able to start to learn to read them.
For example, none of the patterns I have seen signposted on here are "copy and pasteable" but by looking at them I have learnt so much to take to my chosen "design system" to figure out my own solution. 

I turn to The A11y Project because I believe that there is good advice here, not to do the work for me. So, my choice would be keep the patterns, encourage people to add missing patterns, once they go out of date, ask for an update, delete if necessary. 

We could also recommend patterns, as written on blogs and github, and "upvote" them - showing where people are finding a particular pattern still relevant. Those at the bottom of the list may well be because they are badly explained, or just poor / outdated patterns.

**Comment by ericwbailey** [comment]
Just want to keep the ball rolling here. What should we do in the near-term about issues like https://github.com/a11yproject/a11yproject.com/issues/803?

**Comment by scottaohara** [comment]
We could do another pass of cleanup...

**Comment by ericwbailey** [comment]
Patterns have been deprecated.
